custom/icds_reports/static/js/directives/download/download.directive.js

```diff
@@ -345,9 +347,31 @@ function DownloadController($rootScope, $location, locationHierarchy, locationsS
     vm.onSelectYear = function (year) {
         var date = new Date();
         var latest = date;
+        vm.years = vm.yearsCopy;
+        vm.months = vm.monthsCopy;
+
         if (vm.isIncentiveReportSelected()) {
-            var offset = date.getDate() < 15 ? 2 : 1;
-            latest.setMonth(date.getMonth() - offset);
+            vm.years = _.filter(vm.yearsCopy, function (y) {
+                return y.id >= 2018;
+            });
+
+            if (vm.selectedYear === latest.getFullYear()) {
+                var offset = date.getDate() < 15 ? 1 : 0;
+
+                vm.months = _.filter(vm.monthsCopy, function (month) {
+                    return month.id <= (latest.getMonth() + 1) - offset;
+                });
+
+                if (vm.selectedMonth > vm.months[0].id) {
+                    vm.selectedMonth = vm.months[0].id;
+                }
+            } else if (vm.selectedYear === 2018) {
+                vm.months = vm.months.slice(-3);
+                if (vm.selectedMonth < 10) {
+                    vm.selectedMonth = 10;
+                }
+            }
```

Comment:
I think it could be helpful to have this logic encapsulated to a well-named function (maybe even more than one) that explains the intention behind what it's doing so that some future maintainer doesn't have to try and make sense of all this complex date shifting.

Reply:
I was actually referring to the whole code block. E.g. the funkiness with offset triggering based on the 15th as well as the changing of the selected month, etc. all of those are serving a functional purpose that is more comprehensible with words than making sense of the date manipulation, and it would be cleaner if the code read more purposefully. Generally, I think creating well-named functions is a superior practice to comments to achieve that, though can be a cost/benefit for small things...

Reply:
Hi @czue ,
Appreciate your suggestions. I have moved the code block into a separate function. Please have a look and see if it makes sense to you too.

Reply:
Thanks, I think that's a good step. Not sure it's worth going further, though I had in mind that even the little pieces would be split out and/or better documented. For example, I think the offset / month changing logic could be called `setMonthToPreviousIfBeforeThe15th()` or something similar. 

But if @Rohit25negi is happy I'm happy.

Can you fix the whitespace / lint issues?

Reply:
@czue , makes sense. I'll refactor that too along with the lint issues.